drivers/misc/misc_ivshmem_shell.c

```diff
@@ -0,0 +1,175 @@
+		ivshmem = device_get_binding(CONFIG_IVSHMEM_DEV_NAME);
+		if (!ivshmem) {
+			shell_error(shell, "IVshmem device cannot be found\n");
+		}
+	}
+
+	return ivshmem != NULL ? true : false;
+}
+
+static int cmd_ivshmem_shmem(const struct shell *shell,
+			     size_t argc, char **argv)
+{
+	uintptr_t mem;
+	size_t size;
+	uint32_t id;
+	uint16_t vectors;
+
+	if (!get_ivshmem(shell)) {
+		return 0;
+	}
+
+	size = ivshmem_get_mem(ivshmem, &mem);
+	id = ivshmem_get_id(ivshmem);
+	vectors = ivshmem_get_vectors(ivshmem);
+
+	shell_fprintf(shell, SHELL_NORMAL,
+		      "IVshmem up and running: \n"
+		      "\tShared memory: 0x%x of size %u bytes\n"
+		      "\tPeer id: %u\n"
+		      "\tNotification vectors: %u\n",
+		      mem, size, id, vectors);
+
+	return 0;
+}
+
+static int cmd_ivshmem_int(const struct shell *shell,
+			   size_t argc, char **argv)
+{
+	int peer_id;
+	int vector;
+	int ret;
+
+	if (!IS_ENABLED(CONFIG_IVSHMEM_DOORBELL)) {
+		shell_error(shell, "CONFIG_IVSHMEM_DOORBELL is not enabled");
+		return 0;
+	}
+
+	if (!get_ivshmem(shell)) {
+		return 0;
+	}
+
+	peer_id = strtol(argv[1], NULL, 10);
+	vector = strtol(argv[2], NULL, 10);
+
+	ret = ivshmem_int_peer(ivshmem, (uint16_t)peer_id, (uint16_t)vector);
+	if (ret != 0) {
+		shell_error(shell,
+			    "Could not notify peer %u on %u. status %d\n",
+			    peer_id, vector, ret);
+		return -EIO;
+	}
+
+	shell_fprintf(shell, SHELL_NORMAL,
+		      "Notification sent to peer %u on vector %u\n",
+		      peer_id, vector);
+
+	return 0;
+}
+
+static int cmd_ivshmem_get_notified(const struct shell *shell,
+				    size_t argc, char **argv)
+{
+#ifdef CONFIG_IVSHMEM_DOORBELL
+	int vector;
+
+	if (!get_ivshmem(shell)) {
+		return 0;
+	}
+
+	vector = strtol(argv[1], NULL, 10);
+
+	if (ivshmem_register_handler(ivshmem, &doorbell_sig,
+				     (uint16_t)vector)) {
+		shell_error(shell, "Could not get notifications on vector %u",
+			    vector);
+	} else {
+		shell_fprintf(shell, SHELL_NORMAL,
+			      "Notifications enabled for vector %u\n", vector);
+
+		if (!doorbell_started) {
+			k_thread_create(
+				&doorbell_thread,
+				doorbell_stack, STACK_SIZE,
+				(k_thread_entry_t)doorbell_notification_thread,
+				(void *)shell, NULL, NULL,
+				K_PRIO_COOP(7), 0, K_NO_WAIT);
+
+			doorbell_started = true;
+		}
+	}
+#else
+	shell_error(shell, "CONFIG_IVSHMEM_DOORBELL is not enabled");
+#endif
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_ivshmem_cmds,
+			       SHELL_CMD(shmem, NULL,
+					 "Show shared memory info",
+					 cmd_ivshmem_shmem),
+			       SHELL_CMD_ARG(int_peer, NULL,
+					     "Notify a vector on a peer",
+					     cmd_ivshmem_int, 2, 0),
+			       SHELL_CMD_ARG(get_notified, NULL,
+					     "Get notification on vector",
+					     cmd_ivshmem_get_notified, 1, 0),
+			       SHELL_SUBCMD_SET_END
+		);
+
+SHELL_CMD_REGISTER(ivshmem, &sub_ivshmem_cmds,
+		   "IVshmem information", cmd_ivshmem_shmem);
+
+SHELL_CMD_REGISTER(ivshmem_int, &sub_ivshmem_cmds,
```

Comment:
I recommend using the macro `SHELL_CMD_ARG_REGISTER`:

```
SHELL_COND_CMD_ARG_REGISTER(ivshmem_int, &sub_ivshmem_cmds,
"Notify a vector on an ivshmem peer", cmd_ivshmem_int, 3, 0);
```

It will ensure that command is called with the proper number of parameters: `peer_id` and `vector`.

Reply:
true